Add support for Siglent eCal emulation.

jwise opened the pull request:
This patch allows a Siglent SVA1032X to calibrate using a LibreCAL.  It adds support in the firmware to boot into a Siglent-compatible mode (removes USB MSC, to avoid the instrument's UI trying to write to the disk; adds a very limited USB TMC that knows how to speak the Siglent calibration commands) when Siglent calibration coefficients are present, and the unit has the FUNCTION key held while powering on.

We also add a `convert_siglent.py` script, which converts the calibration data stored in the LibreCAL's native Touchstone format to the reverse-engineered header and ZIP+CSV format that a Siglent VNA requires.  Some fields in the Siglent header are not fully understood, but they don't seem to make a difference at least on SVA1032X.

The calibration appears to provide good results tested against other calibration standards I have lying around, and so the S-parameter conversion for ports 1 and 2 at least are probably correct. Unimplemented is an "ATT" mode, used for the VNA's Confidence Check function; I suspect this may be an attenuator, but without a real Siglent eCal, I can't know for sure, and more to the point, the LibreCAL doesn't have one other than the Through port, so we do not implement it.

Here's a video of this operating: https://www.youtube.com/watch?v=H64bFWH6v2k

jankae committed:
switch FW build to Ubuntu 24.04

jwise committed:
Add support for Siglent eCal emulation.

This patch allows a Siglent SVA1032X to calibrate using a LibreCAL.  It
adds support in the firmware to boot into a Siglent-compatible mode
(removes USB MSC, to avoid the instrument's UI trying to write to the
disk; adds a very limited USB TMC that knows how to speak the Siglent
calibration commands) when Siglent calibration coefficients are present,
and the unit has the FUNCTION key held while powering on.

We also add a `convert_siglent.py` script, which converts the
calibration data stored in the LibreCAL's native Touchstone format to
the reverse-engineered header and ZIP+CSV format that a Siglent VNA
requires.  Some fields in the Siglent header are not fully understood,
but they don't seem to make a difference at least on SVA1032X.

The calibration appears to provide good results tested against other
calibration standards I have lying around, and so the S-parameter
conversion for ports 1 and 2 at least are probably correct.
Unimplemented is an "ATT" mode, used for the VNA's Confidence Check
function; I suspect this may be an attenuator, but without a real
Siglent eCal, I can't know for sure, and more to the point, the LibreCAL
doesn't have one other than the Through port, so we do not implement it.

Signed-off-by: Joshua Wise <joshua@joshuawise.com>

miek commented:
> I think someone else (@miek?) tried emulating an Agilent one, too.  Could be nice to have EEPROMs of those, too.

Yep, I did a write-up of some of the initial reverse-engineering & emulation here: https://greatscottgadgets.com/2024/11-13-reverse-engineering-a-vna-ecal-interface-with-cynthion/ and the code & EEPROM dumps are here: https://github.com/miek/ecal-reversing
I would be keen to do some more work on it and integrate support in the LibreCAL firmware if there's interest in that.

For now though, I've been using this script to run the process remotely: https://github.com/miek/vnautils/blob/main/src/vnautils/pna_librecal.py

> What does `ATT` actually do?  Can it be implemented with the same as `THROUGH`, and be a slightly-less-confidence check?

I suspect it selects a second through path with some attenuation (then after the cal it can then be measured and compared with the stored data as a rough verification). It probably could be the same as `THROUGH`, but wouldn't really give any extra confidence at all.

jankae committed:
Build test with older toolchain

jankae committed:
Build test with older toolchain

jankae committed:
Build test with older toolchain

jwise commented:
Interesting.  Not a surprise that auto orientation doesn't work if it uses the same mechanism as 'confidence check'.  Does it send `SL ATT,x,y` commands?

As it turns out SVA1032X doesn't send `*IDN?` at all and I just implemented that at the last minute to make sure I could test it from my host!  So that's lucky that you noticed that.  How are you getting debug data about commands -- do you have a serial port on your LibreCAL?  If so, it would be good to log communications between the SNA5014A and the LibreCAL.

Sometimes in Siglent land, they hardcode things for model numbers.  So you might want to try having your `*IDN?` reply with a 4-port SIglent model number.  Also consider trying different values for the 0x4E and 0x4F bytes in the info.dat (you may have to reboot the VNA to get it to reload the info file from the ecal -- at least, I do on SVA1032X!) -- I have no idea what they do, only that they are copied into the `ecal_model_t` struct specially when it is loaded on SVA1032X.

If that doesn't do it, I guess the next steps would either be to pop the main binary from the SNA5000 into a disassembler, or wait until we can get access to a real eCal to dump the header from that.

jwise commented:
I assume at some point in the past you had a relatively lengthy 'loading data...' period, and it skipped it this time because it still has the correct md5 of zip file cached.

Definitely try *IDN? replying with model numbers (as well as the header data generated by the python script), and also try experimenting with bytes 0x4E and 0x4F.

I think I might have someone who will be able to get a real eCal's data in the next week or so, as well.

jwise commented:
One thing that you might consider is mapping ATT to THROUGH, modifying `outf.write("#HZ,A,B,C,D,T_AB,T_AC,T_AD,T_BC,T_BD,T_CD\n")` to be `outf.write("#HZ,A,B,C,D,T_AB,T_AC,T_AD,T_BC,T_BD,T_CD,CF_AB,CF_AC,CF_AD,CF_BC,CF_BD,CF_CD\n")`, and then repeating the two-port coefficients in the data file (i.e., `for port in ["12", "13", "14", "23", "24", "34"] * 2:`).

jankae committed:
Merge branch 'build_FW_test' into siglent

jankae committed:
Improvements to Siglent emulation mode

- handle *IDN? response and SET:PORT commands send by some VNAs
- default to emulation mode if files are present (press function to force default mode)
- Adjusted script to set number of ports and handle early LibreCALs with partially missing factory data
- Add optional logging to LibreCAL firmware
- indicate emulation mode by blinking wait/ready LEDs

jankae commented:
Merged into the new [siglent](/jankae/LibreCAL/tree/siglent) branch. See fa75511de3f93873b248fce9c4787603267d3b03 for the changes/improvements I made.

This seems to work quite well with my SNA5014A now (as long as I manually assign ports) but I will hold off on merging this into main until I have done more tests (full 4 port calibration and testing whether everything still works with the LibreVNA).

Could you please check whether I broke anything for your SVA1032X? If you don't want to compile it yourself you can also use the artifact [here](https://github.com/jankae/LibreCAL/actions/runs/15346439955/artifacts/3228253091).

jwise committed:
scripts/convert_siglent: make some Pythonic cleanups

Signed-off-by: Joshua Wise <joshua@joshuawise.com>

jwise committed:
sw/main: be more specific in the name of getUsbMode()

Signed-off-by: Joshua Wise <joshua@joshuawise.com>

jwise committed:
sw/siglent_tmc: add support for ATT mode; C style cleanups

Signed-off-by: Joshua Wise <joshua@joshuawise.com>

jwise committed:
sw/usb_descriptors: support CDC in Siglent mode to make it possible to kick the device into bootloader mode without pressing buttons

Signed-off-by: Joshua Wise <joshua@joshuawise.com>

jwise commented:
I added support for ATT (as a clone of THRU, since we don't have an attenuator to mux in) and it seems to make the confidence check thing work.  I wonder if that'll make your auto-assign work?

jankae commented:
Thanks for the review and for cleaning up after me ;)

I'll run some more tests later today and report back

jankae commented:
Merged into the siglent branch again (together with unrelated bug fixes).

After some more testing I think this works absolutely flawlessly:
- Auto orientation detection works
- Full 4-port calibration on the Siglent works
- Calibration with the LibreVNA still works (when in default USB mode)

Well, there is one flaw but that is more on Siglent: My LibreCAL is missing P34_THRU coefficients above 7.9 GHz (this was an early bug in the factory calibration procedure). This means that all other Siglent coefficients above 7.9 GHz are also missing. But the SNA5014A happily calibrates all the way up to 8.5 GHz without any complaints. Calibrated data above 7.9 GHz is obviously bad since it has no coefficients for that. The ECal information screen correctly shows the upper frequency limit of 7.9 GHz.

Unless there are any concerns from your side I will merge this into main and create the 0.3.0 release from it (I think this warrants an increase in the minor version instead of just incrementing to 0.2.4).

jwise commented:
Seems reasonable to me. Consider also whether the GUI should have a button to write a parameter file out while doing the firmware upgrade, and/or whether there should be manual changes?On May 31, 2025, at 11:41, Jan Käberich ***@***.***> wrote:﻿jankae left a comment (jankae/LibreCAL#31)
Merged into the siglent branch again (together with unrelated bug fixes).
After some more testing I think this works absolutely flawlessly:

Auto orientation detection works
Full 4-port calibration on the Siglent works
Calibration with the LibreVNA still works (when in default USB mode)

Well, there is one flaw but that is more on Siglent: My LibreCAL is missing P34_THRU coefficients above 7.9 GHz (this was an early bug in the factory calibration procedure). This means that all other Siglent coefficients above 7.9 GHz are also missing. But the SNA5014A happily calibrates all the way up to 8.5 GHz without any complaints. Calibrated data above 7.9 GHz is obviously bad since it has no coefficients for that. The ECal information screen correctly shows the upper frequency limit of 7.9 GHz.
Unless there are any concerns from your side I will merge this into main and create the 0.3.0 release from it (I think this warrants an increase in the minor version instead of just incrementing to 0.2.4).

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>
